// ebay-promised/lib/definitions/nodes.date.js
"use strict";

// O(1) lookup for Date/Time nodes

module.exports = {
    shippingtime: true,
    starttime: true,
    endtime: true,
    scheduletime: true,
    createdtime: true,
    hardexpirationtime: true,
    invoicedate: true,
    begindate: true,
    enddate: true,
    startcreationtime: true,
    endcreationtime: true,
    endtimefrom: true,
    endtimeto: true,
    updatetime: true,
    lastupdatetime: true,
    lastmodifiedtime: true,
    modtimefrom: true,
    modtimeto: true,
    createtimefrom: true,
    createtimeto: true,
    starttimefrom: true,
    starttimeto: true,
    timeto: true,
    paymenttimefrom: true,
    paymenttimeto: true,
    inventorycountlastcalculateddate: true,
    registrationdate: true,
    timefrom: true,
    timestamp: true,
    messagecreationtime: true,
    resolutiontime: true,
    date: true,
    bankmodifydate: true,
    creditcardexpiration: true,
    creditcardmodifydate: true,
    lastpaymentdate: true,
    submittedtime: true,
    announcementstarttime: true,
    eventtime: true,
    periodicstartdate: true,
    modtime: true,
    expirationtime: true,
    creationtime: true,
    lastusedtime: true,
    disputecreatedtime: true,
    disputemodifiedtime: true,
    externaltransactiontime: true,
    commenttime: true,
    lastbidtime: true,
    time: true,
    creationdate: true,
    lastmodifieddate: true,
    receivedate: true,
    expirationdate: true,
    resolutiondate: true,
    lastreaddate: true,
    userforwarddate: true,
    itemendtime: true,
    userresponsedate: true,
    nextretrytime: true,
    deliverytime: true,
    timebid: true,
    paidtime: true,
    shippedtime: true,
    expectedreleasedate: true,
    paymenttime: true,
    promotionalsalestarttime: true,
    promotionalsaleendtime: true,
    refundtime: true,
    refundrequestedtime: true,
    refundcompletiontime: true,
    estimatedrefundcompletiontime: true,
    lastemailsenttime: true,
    sellerinvoicetime: true,
    estimateddeliverydate: true,
    printedtime: true,
    deliverydate: true,
    refundgrantedtime: true,
    scheduleddeliverytimemin: true,
    scheduleddeliverytimemax: true,
    actualdeliverytime: true,
    usebydate: true,
    lastopenedtime: true,
    returndate: true,
    revocationtime: true,
    lasttimemodified: true,
    createddate: true,
    invoicesenttime: true,
    acceptedtime: true,
    sellerebaypaymentprocessenabletime: true,
    useridlastchanged: true,
    actionrequiredby: true
};
