// ebay-promised/lib/Parser.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _bluebird = require("bluebird");

var _bluebird2 = _interopRequireDefault(_bluebird);

var _ecjson = require("ecjson");

var _ecjson2 = _interopRequireDefault(_ecjson);

var _errors = require("./errors");

var _Immutable = require("./utils/Immutable");

var _Immutable2 = _interopRequireDefault(_Immutable);

var _extraneous = require("./definitions/extraneous");

var _extraneous2 = _interopRequireDefault(_extraneous);

var _nodes = require("./definitions/nodes.date");

var _nodes2 = _interopRequireDefault(_nodes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var ITERABLE_KEY = /Array|List/;

/**
 * A collection of pure methods that are used to parse eBay API responses
 * should generally be bound to a Request via:
 *   `Function.prototype.bind`
 *   `Promise.prototype.bind`
 *  
 */

var Parser = function () {
  function Parser() {
    _classCallCheck(this, Parser);
  }

  _createClass(Parser, null, [{
    key: "toJSON",


    /**
     * converts an XML response to JSON
     *
     * @param      {XML}     xml     The xml
     * @return     {Promise}         resolves to a JSON representation of the HTML 
     */
    value: function toJSON(xml) {
      return new _bluebird2.default(function (resolve, reject) {
        _ecjson2.default.XmlToJson(xml, resolve);
      });
    }

    /**
     * unwraps a verb Response from eBay
     * must be verbed within the context of an {Ebay.Response}
     *
     * @param      {Call}    verb    The verb
     * @return     {Object}          The unwrapped verb
     */

  }, {
    key: "unwrap",
    value: function unwrap(req, json) {
      return Parser.flatten(json[req.responseWrapper]);
    }

    /**
     * Casts text representations to Javascript representations
     *
     * @param      {String}       value   The value
     * @return     {Date|Number}          The cast value
     */

  }, {
    key: "cast",
    value: function cast(value, key) {

      if (!isNaN(value)) return Number(value);

      if (value === "true") return true;

      if (value === "false") return false;

      if (typeof key === 'string' && _nodes2.default[key.toLowerCase()]) {
        return new Date(value);
      }

      return value;
    }

    /**
     * recursively flattens `value` keys in the XML -> JSON conversion
     * we can do this because we don't need to worry about XML attributes from eBay
     *
     * @param      {Object}  o       the object output from the XML parser
     * @return     {Object}          the flattened output
     */

  }, {
    key: "flatten",
    value: function flatten(o, key) {

      if (o.value) {
        return Parser.cast(o.value, key);
      }

      if (Array.isArray(o)) {
        return o.map(Parser.flatten);
      }

      if ((typeof o === "undefined" ? "undefined" : _typeof(o)) !== "object") {
        return Parser.cast(o, key);
      }

      return Object.keys(o).reduce(function (deflated, key) {
        deflated[key] = Parser.flatten(o[key], key);
        return deflated;
      }, {});
    }

    /**
     * flattens the eBay pagination object to be easier to deal with
     *
     * @param      {Object}  obj    the JSON representation of a Response
     * @return     {Object}         the friendly pagination extended Response
     */

  }, {
    key: "parsePagination",
    value: function parsePagination(obj) {
      if (!obj.PaginationResult) return {};

      var p = obj.PaginationResult;
      delete obj.PaginationResult;

      return { pagination: {
          pages: p.TotalNumberOfPages || 0,
          length: p.TotalNumberOfEntries || 0
        } };
    }
    /**
     * cleans the Ebay response up
     *
     * @param      {Object}  res     The response object
     * @return     {Object}  res     The cleaned response
     */

  }, {
    key: "clean",
    value: function clean(res) {

      if (res.Ack === "Error" || res.Ack === "Failure") {
        _errors.throws.Ebay_Api_Error(res.Errors);
      }

      // Drop extraneous keys
      res = Object.keys(res).filter(function (key) {
        return !~_extraneous2.default.indexOf(key);
      }).reduce(function (acc, key) {
        acc[key] = res[key];
        return acc;
      }, {});

      return Parser.fold(res);
    }

    /**
     * recursively folds a response from eBay into something reasonable
     *
     * @param      {Object}  res     The resource
     * @return     {Object}          The folded response
     */

  }, {
    key: "fold",
    value: function fold(res) {
      return Object.keys(res).reduce(function (cleaned, key) {
        var value = res[key];
        if (key.match(/List/)) {
          return _Immutable2.default.merge(cleaned, Parser.parsePagination(value), Parser.getList(value));
        }

        if (key.match(/Array/)) {
          return _Immutable2.default.merge(cleaned, Parser.getList(value));
        }

        cleaned[key] = value;
        return cleaned;
      }, {});
    }

    /**
     * Gets the List element from an eBay response
     *
     * @param      {<type>}  list    The list
     * @return     {Object}          The list.
     */

  }, {
    key: "getList",
    value: function getList(list) {
      var parent = Parser.getMatchingKey(list, "Array");
      var wrapper = Object.keys(parent)[0];
      var entries = parent[wrapper] || [];
      // Ensure we always have an Iterable interface for things that should be iterable
      return { results: Array.isArray(entries) ? entries : [entries] };
    }

    /**
     * Gets the matching key.
     *
     * @param      {<type>}  obj     The object
     * @param      {<type>}  substr  The substr to match
     * @return     {<type>}          The matching key.
     */

  }, {
    key: "getMatchingKey",
    value: function getMatchingKey(obj, substr) {
      var keys = Object.keys(obj);
      while (keys.length) {
        var key = keys.pop();
        if (~key.indexOf(substr)) return obj[key];
      }
      return obj;
    }
  }]);

  return Parser;
}();

exports.default = Parser;
