// ebay-promised/lib/errors/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.throws = exports.Ebay_Api_Error = exports.Env_Error = exports.Setting_Error = exports.No_Call_Error = exports.No_Auth_Token_Error = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _es6Error = require("es6-error");

var _es6Error2 = _interopRequireDefault(_es6Error);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Error object for ease of capturing if some service depends on .toJSON() method to log something
 * 
 * @ignore
 */
var Error = function (_Extendable_Error) {
  _inherits(Error, _Extendable_Error);

  function Error() {
    _classCallCheck(this, Error);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(Error).apply(this, arguments));
  }

  _createClass(Error, [{
    key: "toJSON",


    /**
     * returns a JSON representation of the Error
     *
     * @return     {Object}  json representation of the Error
     */
    value: function toJSON() {
      return {
        message: this.message,
        stack: this.stack,
        type: this.constructor.name,
        meta: this.meta || null
      };
    }
  }]);

  return Error;
}(_es6Error2.default);

/**
 * thrown when Request.prototype.run() is called without an authToken
 *
 * @class      No_Auth_Token (name)
 */


var No_Auth_Token_Error = exports.No_Auth_Token_Error = function (_Error) {
  _inherits(No_Auth_Token_Error, _Error);

  function No_Auth_Token_Error() {
    var msg = arguments.length <= 0 || arguments[0] === undefined ? "no authToken present.  Please invoke `Ebay.prototype.authToken(<Token>)`." : arguments[0];

    _classCallCheck(this, No_Auth_Token_Error);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(No_Auth_Token_Error).call(this, msg));
  }

  return No_Auth_Token_Error;
}(Error);

/**
 * thrown when Request.prototype.run() is called without having defined an eBay API call
 */

var No_Call_Error = exports.No_Call_Error = function (_Error2) {
  _inherits(No_Call_Error, _Error2);

  function No_Call_Error() {
    var msg = arguments.length <= 0 || arguments[0] === undefined ? "no eBay API call defined, please invoke one." : arguments[0];

    _classCallCheck(this, No_Call_Error);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(No_Call_Error).call(this, msg));
  }

  return No_Call_Error;
}(Error);

/**
 * thrown when attempting to change a setting that cannot be changed
 *
 * @param {String} setting    the setting that was attempted
 */

var Setting_Error = exports.Setting_Error = function (_Error3) {
  _inherits(Setting_Error, _Error3);

  function Setting_Error(setting) {
    _classCallCheck(this, Setting_Error);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(Setting_Error).call(this, "cannot configure \"state." + setting + "\" at this time, are you trying to define a Global on a Request?"));
  }

  return Setting_Error;
}(Error);

/**
 * thrown when attempting to load environment variables that don't exist
 */


var Env_Error = exports.Env_Error = function (_Error4) {
  _inherits(Env_Error, _Error4);

  function Env_Error(key) {
    _classCallCheck(this, Env_Error);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(Env_Error).call(this, "could not find " + key + " in process.env"));
  }

  return Env_Error;
}(Error);

/**
 * Thrown when an Error occurs on eBay's side.
 * Allows for easier control flow with Promises
 * 
 * @example
 * 
 * Ebay
 *  .create(config)
 *  .GetSellerList()
 *  .run()
 *  .catch(errors.Ebay.Api_Error, App.handleErr)
 *  .catch( function (other_error) {
 *    handle(other_error)
 *  })
 */


var Ebay_Api_Error = exports.Ebay_Api_Error = function (_Error5) {
  _inherits(Ebay_Api_Error, _Error5);

  function Ebay_Api_Error(err) {
    _classCallCheck(this, Ebay_Api_Error);

    var _this6 = _possibleConstructorReturn(this, Object.getPrototypeOf(Ebay_Api_Error).call(this, err.LongMessage || err.ShortMessage));

    _this6.meta = err;
    return _this6;
  }

  return Ebay_Api_Error;
}(Error);

/**
 * convenience methods for Error creation
 * 
 * @ignore
 * @type {Object}
 * 
 * @example
 *  throws.Error("an error message")
 *  throws.No_Auth_Token()
 */


var throws = exports.throws = Object.keys(exports).reduce(function (thrower, err) {
  if (err === "throws") return thrower;
  var cstr = exports[err];
  thrower[err] = function _thrower() {
    throw new (Function.prototype.bind.apply(cstr, [null].concat(Array.prototype.slice.call(arguments))))();
  };
  return thrower;
}, {});
